### PR TITLE
Ftx update ratelimiter

### PR DESCRIPTION
**Pull request by DoctorSlimm**
https://help.ftx.com/hc/en-us/articles/360052595091-Ratelimits-on-FTX
https://docs.ftx.com/#rate-limits

- pretty confusing docs, not sure what the '10 requests a second' refers to as opposed to the 6 requests a second whilst 'only creating orders counts towards ratelimit'
- creating, cancelling, retrieving orders tested and polled all good
- though when looping creating orders and cancelling them I do believe their docs are correct in stating a significant increase in latency despite not necessarily being banned and receiving HTTP 429 errors
- other endpoints such as fetching balances also polled and work same as marketdata endpoints

### CODE REVIEW

**Comment by frosty00** [comment]
let's change the 0 ratelimit endpoints, that will certainly get you banned

**Comment by frosty00** [comment]
Screenshot 2022-01-10 at 01 41 25

instead of doing this let's just change the number 100 to 33.3 and keep all the weights as 1

**Comment by DoctorSlimm** [comment]
@frosty00 Okay sure thing, you mean change all the zero weights to 1 and keep the creating order endpoints as they are? the zero weights are for endpoints which 'dont cost the ratelimit' - the docs say in small letters than market data endpoints dont cost and its only the creating order ones that do

**Comment by DoctorSlimm** [comment]
@frosty00 so even though they say in the docs that cancels and marketdata endpoints do not contribute to the ratelimiter -  I should still set the rateLimit property to 33.3 and set all weights for all endpoints to be 1 regardless of wether its an order making or marketdata endpoint?

**Comment by kroitor** [comment]
> instead of doing this let's just change the number 100 to 33.3 and keep all the weights as 1

I would suggest to adjust the weights in such a manner that we have our cost equal to their cost. That would require tweaking our rateLimit value.

**Comment by DoctorSlimm** [comment]
@kroitor and what about non-costing endpoints? or do we think it will be hard to decide which they may consider to be market data or not - and we may end up getting banned because misjudged an endpoint so better to just ratelimit all the endpoints?

**Comment by kroitor** [comment]
@DoctorSlimm 

> @kroitor and what about non-costing endpoints? 

Technically, our cost can't be 0, because that will make the rate limiter fire unlimited requests.

> or do we think it will be hard to decide which they may consider to be market data or not - and we may end up getting banned because misjudged an endpoint so better to just ratelimit all the endpoints?

We should set at least some minimal rate limit on those zero-cost endpoints, I don't think FTX will like it if we send 3 million market data requests per second, despite that they say it costs nothing.

**Comment by DoctorSlimm** [comment]
@frosty00 @kroitor I set the rateLimit to 33.34 because technically the hard limit is 33.333333ms between requests, so just in case of rounding errors to avoid bans

**Comment by frosty00** [comment]
dw that one hundredth of a milliseconds will probably be wasted during the code execution anyway

File: js/ftx.js [hunk 1]
```diff
@@ -93,179 +97,179 @@ module.exports = class ftx extends Exchange {
...
+                        'subaccounts': 1,
+                        'subaccounts/update_name': 1,
+                        'subaccounts/transfer': 1,
                         // account
-                        'account/leverage',
+                        'account/leverage': 1,
                         // wallet
-                        'wallet/withdrawals',
-                        'wallet/saved_addresses',
+                        'wallet/withdrawals': 90.1,
```
Comment (kroitor) [inline js/ftx.js:222]: I think we can set this to 90 now?

**Comment by kroitor** [comment]
@DoctorSlimm a quick question – you tested it by polling their public endpoints for a while and they didn't ban you, right?

**Comment by DoctorSlimm** [comment]
@kroitor correct, and I also spammed the createOrder Endpoint, and also a creating-then cancelling loop

**Pull request merged**